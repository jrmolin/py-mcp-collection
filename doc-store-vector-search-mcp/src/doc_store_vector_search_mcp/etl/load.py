from abc import ABC, abstractmethod
from collections.abc import AsyncGenerator
from typing import Any

from langchain_community.document_loaders import DirectoryLoader as LangchainDirectoryLoader
from langchain_community.document_loaders import RecursiveUrlLoader as LangchainRecursiveUrlLoader
from langchain_community.document_loaders import WebBaseLoader as LangchainWebBaseLoader
from langchain_community.document_loaders.json_loader import JSONLoader as LangchainJSONLoader
from langchain_community.document_loaders.sitemap import SitemapLoader as LangchainSitemapLoader
from langchain_core.documents import Document


type LoaderTypes = WebPageLoader | SiteMapLoader | RecursiveWebLoader | DirectoryLoader | JSONJQLoader


class WebPageLoader:

    @classmethod
    async def load(cls, url: str, **kwargs: Any) -> AsyncGenerator[list[Document], None]:
        loader = LangchainWebBaseLoader(url, **kwargs)

        async for document in loader.alazy_load():
            yield document


class SiteMapLoader:

    @classmethod
    async def load(cls, url: str, filter_urls: list[str] | None = None, **kwargs: Any) -> AsyncGenerator[list[Document], None]:
        loader = LangchainSitemapLoader(url, filter_urls=filter_urls, **kwargs)

        async for document in loader.alazy_load():
            yield document


class RecursiveWebLoader:

    async def load(cls, url: str, **kwargs: Any) -> AsyncGenerator[list[Document], None]:
        loader = LangchainRecursiveUrlLoader(url, use_async=True, **kwargs)

        async for document in loader.alazy_load():
            yield document


class DirectoryLoader:

    @classmethod
    async def load(
        cls, directory_path: str, glob: str = "**/*.{md,mdx,txt}", **kwargs: Any
    ) -> AsyncGenerator[list[Document], None]:
        text_loader_kwargs = {"autodetect_encoding": True}
        loader = LangchainDirectoryLoader(
            directory_path, glob=glob, silent_errors=False, loader_kwargs=text_loader_kwargs, **kwargs
        )

        async for document in loader.alazy_load():
            yield document


class JSONJQLoader:

    @classmethod
    def load(cls, json_path: str, jq_schema: str = ".items[]", content_key: str = "text", **kwargs: Any) -> AsyncGenerator[list[Document], None]:
        loader = LangchainJSONLoader(json_path, jq_schema=jq_schema, content_key=content_key, **kwargs)

        yield from loader.lazy_load()
