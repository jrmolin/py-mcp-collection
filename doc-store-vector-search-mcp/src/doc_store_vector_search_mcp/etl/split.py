from abc import ABC, abstractmethod
from typing import Any, TypeAlias

import langchain_text_splitters
from langchain_core.documents import Document
from langchain_text_splitters import (
    MarkdownHeaderTextSplitter,
    RecursiveCharacterTextSplitter,
    RecursiveJsonSplitter,
)
from typing_extensions import TypeVar

T = TypeVar("T")


class Splitter(ABC):
    @abstractmethod
    def split_to_documents(self, to_split: T) -> list[Document]:
        pass


CodeSplitterLanguage: TypeAlias = langchain_text_splitters.Language  # noqa: UP040


def tag_document_order(documents: list[Document]) -> list[Document]:
    for i, document in enumerate(documents):
        document.metadata["order"] = i
    return documents


class JsonSplitter(Splitter):
    def __init__(self):
        self.splitter = RecursiveJsonSplitter(
            max_chunk_size=500,
            min_chunk_size=25,
        )

    def split_to_documents(self, to_split: dict[str, Any]) -> list[Document]:
        return tag_document_order(self.splitter.create_documents([to_split], convert_lists=True))


class MarkdownSplitter:
    def __init__(self):
        self.header_splitter = MarkdownHeaderTextSplitter(
            headers_to_split_on=[
                ("#", "Header 1"),
                ("##", "Header 2"),
            ],
            strip_headers=False,
        )
        self.chunk_splitter = RecursiveCharacterTextSplitter(
            chunk_size=300,
            chunk_overlap=25,
        )

    def split_to_documents(self, text: str) -> list[Document]:
        per_header_documents: list[Document] = self.header_splitter.split_text(text)
        chunked_documents: list[Document] = []

        for document in per_header_documents:
            chunked_documents.extend(self.chunk_splitter.split_documents([document]))

        return tag_document_order(chunked_documents)


class CodeSplitter:
    def __init__(self, language: CodeSplitterLanguage):
        self.splitter = RecursiveCharacterTextSplitter.from_language(
            language,
            chunk_size=300,
            chunk_overlap=25,
        )

    def split_to_documents(self, text: str) -> list[Document]:
        chunks = self.splitter.split_text(text)
        return tag_document_order(self.splitter.create_documents(chunks))
