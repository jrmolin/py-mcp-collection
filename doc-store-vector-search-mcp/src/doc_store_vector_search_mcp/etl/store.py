from typing import Any, TypeVar

import duckdb
import torch
from langchain_community.vectorstores import DuckDB, ElasticsearchStore
from langchain_core.documents import Document
from langchain_core.embeddings import Embeddings
from pydantic import BaseModel
from sentence_transformers import SentenceTransformer
from transformers import AutoModel, AutoTokenizer

type VectorStoreTypes = DuckDB | ElasticsearchStore


class ElasticsearchStoreSettings(BaseModel):
    es_url: str
    es_index: str

    def connect(self) -> None:
        pass


class DuckDBSettings(BaseModel):
    db_path: str

    def connect(self) -> duckdb.DuckDBPyConnection:
        return duckdb.connect(database=self.db_path, config={"enable_external_access": "false"})


type VectorStoreSettings = ElasticsearchStoreSettings | DuckDBSettings

T = TypeVar("T", bound=VectorStoreTypes)
S = TypeVar("S", bound=VectorStoreSettings)


class TextEmbedding(Embeddings):
    def __init__(self, model: str = "all-MiniLM-L6-v2"):
        self.model = SentenceTransformer(model)

    def embed(self, text: str) -> list[float]:
        return self.model.encode([text])[0].tolist()

    def embed_documents(self, texts: list[str]) -> list[list[float]]:
        return self.model.encode(texts).tolist()

    def embed_query(self, text: str) -> list[float]:
        return self.model.encode([text])[0].tolist()


class CodeEmbedding(Embeddings):
    def __init__(self):
        self.device = "cpu"

        self.tokenizer = AutoTokenizer.from_pretrained("Salesforce/codet5p-110m-embedding", trust_remote_code=True)
        self.model = AutoModel.from_pretrained("Salesforce/codet5p-110m-embedding", trust_remote_code=True).to(self.device)

        self.torch = torch

    def embed(self, text: str) -> list[float]:
        inputs = self.tokenizer.encode(text, return_tensors="pt").to(self.device)
        with self.torch.no_grad():
            embedding = self.model(inputs)[0]
        return embedding.cpu().numpy().tolist()

    def embed_documents(self, texts: list[str]) -> list[list[float]]:
        return [self.embed(text) for text in texts]

    def embed_query(self, text: str) -> list[float]:
        return self.embed(text)


class VectorStoreManager[T: VectorStoreTypes, S: VectorStoreSettings]:
    def __init__(self, project_name: str, kb_id: str, code_settings: S, document_settings: S, vector_store_class: type[T]):
        self.project_name = project_name
        self.kb_id = kb_id

        self.code_embedding = CodeEmbedding()
        self.document_embedding = TextEmbedding()

        self.code_vector_store: T = vector_store_class(embedding=self.code_embedding, connection=code_settings.connect())  # type: ignore
        self.document_vector_store: T = vector_store_class(embedding=self.document_embedding, connection=document_settings.connect())  # type: ignore

    async def add_code_documents(self, documents: list[Document], metadata: dict[str, Any]):
        [document.metadata.update(metadata or {}) for document in documents]

        await self.code_vector_store.aadd_documents(documents)

    async def add_documents(self, documents: list[Document], metadata: dict[str, Any]):
        [document.metadata.update(metadata or {}) for document in documents]
        await self.document_vector_store.aadd_documents(documents)


class ProjectVectorStoreManager[T: VectorStoreTypes, S: VectorStoreSettings]:
    def __init__(self, project_name: str, code_vector_store: VectorStoreManager[T, S], document_vector_store: VectorStoreManager[T, S]):
        self.project_name = project_name
        self.code_vector_store = code_vector_store
        self.document_vector_store = document_vector_store

    async def add_code_documents(self, documents: list[Document], metadata: dict[str, Any]):
        updated_metadata = {
            **metadata,
            "project_name": self.project_name,
        }
        await self.code_vector_store.add_code_documents(documents, updated_metadata)

    async def add_markdown_documents(self, documents: list[Document], metadata: dict[str, Any]):
        updated_metadata = {
            **metadata,
            "project_name": self.project_name,
        }

        await self.document_vector_store.add_documents(documents, updated_metadata)


class KnowledgeBaseVectorStoreManager:
    def __init__(self, kb_id: str, project_vector_store: ProjectVectorStoreManager):
        self.kb_id = kb_id
        self.project_vector_store = project_vector_store

    async def add_code_documents(self, documents: list[Document], metadata: dict[str, Any]):
        updated_metadata = {
            **metadata,
            "kb_id": self.kb_id,
        }
        await self.project_vector_store.add_code_documents(documents, updated_metadata)

    async def add_markdown_documents(self, documents: list[Document], metadata: dict[str, Any]):
        updated_metadata = {
            **metadata,
            "kb_id": self.kb_id,
        }

        await self.project_vector_store.add_markdown_documents(documents, updated_metadata)

    @classmethod
    def from_project_vector_store(cls, kb_id: str, project_vector_store: ProjectVectorStoreManager) -> "KnowledgeBaseVectorStoreManager":
        return cls(
            kb_id=kb_id,
            project_vector_store=project_vector_store,
        )
