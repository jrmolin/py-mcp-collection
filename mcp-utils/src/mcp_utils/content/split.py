from mcp.types import TextContent


def split_text_content(
    content: TextContent,
    split_size: int,
) -> list[TextContent]:
    """Split a text content into a list of text contents, each of which is less than or equal to the split size."""

    if len(content.text) <= split_size:
        return [content]

    content_text = content.text
    contents = []

    while len(content_text) > split_size:
        contents.append(TextContent(text=content_text[:split_size], type="text"))
        content_text = content_text[split_size:]

    if len(content_text) > 0:
        contents.append(TextContent(text=content_text, type="text"))

    return contents
