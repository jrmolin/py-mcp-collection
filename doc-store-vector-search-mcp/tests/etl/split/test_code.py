import pytest
from langchain_core.documents import Document
from syrupy.assertion import SnapshotAssertion

from doc_store_vector_search_mcp.etl.split import (
    CodeSplitter,
    CodeSplitterLanguage,
)

MOCK_PYTHON_CODE = """
import os

class Manager:
    '''Docstring for Manager class'''
    def __init__(self, name: str):
        self.name = name

    def __str__(self):
        return f"Manager(name={self.name})"

class SubManager(Manager):
    def __init__(self, name: str, age: int):
        super().__init__(name)
        self.age = age

    def __str__(self):
        return f"SubManager(name={self.name}, age={self.age})"

def main():
    manager = Manager("John")
    print(manager)

    subManager = SubManager(manager, 30)
    print(subManager)

if __name__ == "__main__":
    main()
"""

MOCK_GOLANG_CODE = """
package main

import "fmt"

type Manager struct {
    name string
}

type SubManager struct {
    Manager
    age int
}

func (m *Manager) String() string {
    return fmt.Sprintf("Manager(name=%s)", m.name)
}

func (m *SubManager) String() string {
    return fmt.Sprintf("SubManager(name=%s, age=%d)", m.name, m.age)
}

func main() {
    manager := Manager{name: "John"}
    fmt.Println(manager)

    subManager := SubManager{Manager: manager, age: 30}
    fmt.Println(subManager)
}
"""


@pytest.mark.parametrize(
    ("code", "language"),
    [
        (MOCK_PYTHON_CODE, CodeSplitterLanguage.PYTHON),
        (MOCK_GOLANG_CODE, CodeSplitterLanguage.GO),
    ],
    ids=["python", "golang"],
)
def test_codesplitter_splits_by_language_syntax(code: str, language: CodeSplitterLanguage, snapshot: SnapshotAssertion):
    splitter = CodeSplitter(language=language)

    split_documents = splitter.split_to_documents(code)

    assert isinstance(split_documents, list)
    assert len(split_documents) > 0
    for doc in split_documents:
        assert isinstance(doc, Document)

    snapshot.assert_match(split_documents)
