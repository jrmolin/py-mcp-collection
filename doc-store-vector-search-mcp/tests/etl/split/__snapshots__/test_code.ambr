# serializer version: 1
# name: test_codesplitter_splits_by_language_syntax[golang]
  list([
    Document(metadata={'order': 0}, page_content='package main\n\nimport "fmt"\n\ntype Manager struct {\n    name string\n}\n\ntype SubManager struct {\n    Manager\n    age int\n}\n\nfunc (m *Manager) String() string {\n    return fmt.Sprintf("Manager(name=%s)", m.name)\n}'),
    Document(metadata={'order': 1}, page_content='func (m *SubManager) String() string {\n    return fmt.Sprintf("SubManager(name=%s, age=%d)", m.name, m.age)\n}\n\nfunc main() {\n    manager := Manager{name: "John"}\n    fmt.Println(manager)\n\n    subManager := SubManager{Manager: manager, age: 30}\n    fmt.Println(subManager)\n}'),
  ])
# ---
# name: test_codesplitter_splits_by_language_syntax[python]
  list([
    Document(metadata={'order': 0}, page_content='import os\n\nclass Manager:\n    \'\'\'Docstring for Manager class\'\'\'\n    def __init__(self, name: str):\n        self.name = name\n\n    def __str__(self):\n        return f"Manager(name={self.name})"'),
    Document(metadata={'order': 1}, page_content='class SubManager(Manager):\n    def __init__(self, name: str, age: int):\n        super().__init__(name)\n        self.age = age\n\n    def __str__(self):\n        return f"SubManager(name={self.name}, age={self.age})"'),
    Document(metadata={'order': 2}, page_content='def main():\n    manager = Manager("John")\n    print(manager)\n\n    subManager = SubManager(manager, 30)\n    print(subManager)\n\nif __name__ == "__main__":\n    main()'),
  ])
# ---
